node/peers_base.js

```diff
@@ -0,0 +1,122 @@
+// Copyright (c) 2015 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+'use strict';
+
+var inherits = require('util').inherits;
+var EventEmitter = require('./lib/event_emitter');
+
+function TChannelPeersBase(channel, options) {
+    var self = this;
+    EventEmitter.call(self);
+
+    self.channel = channel;
+    self.logger = self.channel.logger;
+    self.options = options || {};
+    self._map = Object.create(null);
+    self._keys = [];
+}
+
+inherits(TChannelPeersBase, EventEmitter);
+
+TChannelPeersBase.prototype.close = function close(peers, callback) {
+    var self = this;
+
+    var counter = peers.length + 1;
+    peers.forEach(function eachPeer(peer) {
+        peer.close(onClose);
+    });
+    self.clear();
+    onClose();
+
+    function onClose() {
+        if (--counter <= 0) {
```

Comment:
Suggest making the counter decrement a statement and flattening the following conditional.